File: c#/BigEgg.Algorithm/BigEgg.Algorithm/Collections/ResizingArrayStack.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace BigEgg.Algorithm.Collections
{
    public class ResizingArrayStack<Item> : IStack<Item>
    {
        private Item[] a;
        private int N;

        /// <summary>
        /// Initializes a new instance of the <see cref="ResizingArrayStack{Item}"/> class.
        /// </summary>
        public ResizingArrayStack()
        {
            N = 0;
            a = new Item[2];
        }

        /// <summary>
        /// Is this stack empty?
        /// </summary>
        /// <returns><c>True</c> if this stack is empty; <c>false</c> otherwise</returns>
        public bool IsEmpty()
        {
            return N == 0;
        }

        /// <summary>
        /// Returns the number of items in the stack.
        /// </summary>
        /// <returns>The number of items in the stack.</returns>
        public int Size()
        {
            return N;
        }

        /// <summary>
        /// Resize the underlying array holding the elements.
        /// </summary>
        /// <param name="capacity">The new capacity of the stack.</param>
        private void Resize(int capacity)
        {
            Item[] temp = new Item[capacity];
            for (int i = 0; i < N; i++)
            {
                temp[i] = a[i];
            }
            a = temp;
        }

        /// <summary>
        /// Adds the item to this stack.
        /// </summary>
        /// <param name="item">The item to add.</param>
        /// <exception cref="System.ArgumentNullException">Item cannot be null.</exception>
        public void Push(Item item)
        {
            if (item == null) { throw new ArgumentNullException("item"); }
            if (N == a.Length) Resize(a.Length * 2);
            a[N++] = item;
        }

        /// <summary>
        /// Removes and returns the item most recently added to this stack.
        /// </summary>
        /// <returns>The item most recently added.</returns>
        /// <exception cref="System.ArgumentOutOfRangeException">Stack is empty.</exception>
        public Item Pop()
        {
            if (IsEmpty()) { throw new ArgumentOutOfRangeException(); }
            var result = a[N - 1];
            a[N - 1] = default(Item);
            N--;
            if (N == a.Length / 4) Resize(a.Length / 2);
            return result;
        }

        /// <summary>
        /// Returns (but does not remove) the item most recently added to this stack.
        /// </summary>
        /// <returns>The item most recently added to this stack.</returns>
        /// <exception cref="System.ArgumentOutOfRangeException">Stack is empty.</exception>
        public Item Peek()
        {
            if (IsEmpty()) { throw new ArgumentOutOfRangeException(); }
            return a[N - 1];
        }

        /// <summary>
        /// Returns a string representation of this stack.
        /// </summary>
        /// <returns>The sequence of items in the stack in LIFO order, separated by spaces.</returns>
        public override string ToString()
        {
            return String.Join(" ", this);
        }

        /// <summary>
        /// Returns an enumerator that iterates through the items in LIFO order.
        /// </summary>
        /// <returns>
        /// A <see cref="T:System.Collections.Generic.IEnumerator`1" /> that can be used to iterate through the items in LIFO order.
        /// </returns>
        public IEnumerator<Item> GetEnumerator()
        {
            return new Enumerator(this);
        }

        /// <summary>
        /// Returns an enumerator that iterates through a items in LIFO order.
        /// </summary>
        /// <returns>
        /// An <see cref="T:System.Collections.IEnumerator" /> object that can be used to iterate through the items in LIFO order.
        /// </returns>
        IEnumerator IEnumerable.GetEnumerator()
        {
            return new Enumerator(this);
        }

        private struct Enumerator : IEnumerator<Item>, IDisposable, IEnumerator
        {
            private Item[] a;
            private int i;
            private int N;

            public Enumerator(ResizingArrayStack<Item> stack)
            {
                a = stack.a;
                i = stack.N;
                N = stack.N;
            }


            public Item Current
            {
                get { return a[i - 1]; }
            }

            public void Dispose()
            {
                a = null;
                N = 0;
            }

            object System.Collections.IEnumerator.Current
            {
                get { return a[i - 1]; }
            }

            public bool MoveNext()
            {
                i--;
                return i > 0;
            }

            public void Reset()
            {
                i = N;
            }
        }
    }
}

File: java/Algorithms/src/Collections/ResizingArrayStack.java
package Collections;

import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * Created by jianming.xiao on 10/7/14.
 */
public class ResizingArrayStack<Item> implements IStack<Item> {
    private Item[] a;
    private int N;

    /**
     * Initializes an empty stack.
     */
    public ResizingArrayStack() {
        a = (Item[]) new Object[2];
        N = 0;
    }

    /**
     * Is this stack empty?
     *
     * @return true if this stack is empty; false otherwise
     */
    @Override
    public boolean isEmpty() {
        return N == 0;
    }

    /**
     * Returns the number of items in the stack.
     *
     * @return the number of items in the stack
     */
    @Override
    public int size() {
        return N;
    }

    private void resize(int capacity) {
        Item[] temp = (Item[]) new Object[capacity];
        for (int i = 0; i < N; i++) {
            temp[i] = a[i];
        }
        a = temp;
    }

    /**
     * Adds the item to this stack.
     *
     * @param item the item to add
     * @throws NullPointerException if the <tt>item</tt> is null;
     */
    @Override
    public void push(Item item) throws NullPointerException {
        if (item == null) throw new NullPointerException();
        if (N == a.length) resize(2 * a.length);
        a[N++] = item;
    }

    /**
     * Removes and returns the item most recently added to this stack.
     *
     * @return the item most recently added
     * @throws java.util.NoSuchElementException if this stack is empty
     */
    @Override
    public Item pop() throws NoSuchElementException {
        if (isEmpty()) throw new NoSuchElementException("Stack underflow");
        Item result = a[N - 1];
        a[N - 1] = null;
        N--;
        if (N == a.length / 4) resize(a.length / 2);
        return result;
    }

    /**
     * Returns (but does not remove) the item most recently added to this stack.
     *
     * @return the item most recently added to this stack
     * @throws java.util.NoSuchElementException if this stack is empty
     */
    @Override
    public Item peek() throws NoSuchElementException {
        if (isEmpty()) throw new NoSuchElementException("Stack underflow");
        return a[N - 1];
    }

    /**
     * Returns a string representation of this stack.
     *
     * @return the sequence of items in the stack in LIFO order, separated by spaces
     */
    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();
        for (Item item : this) {
            stringBuilder.append(item.toString()).append(" ");
        }
        return stringBuilder.toString();
    }

    /**
     * Returns an iterator to this stack that iterates through the items in LIFO order.
     *
     * @return an iterator to this stack that iterates through the items in LIFO order.
     */
    @Override
    public Iterator<Item> iterator() {
        return new ReverseArrayIterator();
    }

    private class ReverseArrayIterator implements Iterator<Item> {
        private int i = N;

        @Override
        public boolean hasNext() {
            return i > 0;
        }

        @Override
        public Item next() {
            if (!hasNext()) throw new NoSuchElementException();
            return a[--i];
        }

        @Override
        public void remove() {
            throw new UnsupportedOperationException();
        }
    }
}
